### PR TITLE
Feature/website message

### PR DESCRIPTION
**Pull request by wzh425**

### CODE REVIEW

**Review by doddgu — APPROVED** [review]
.

File: src/Contracts/Masa.Mc.Contracts.Admin/Dtos/WebsiteMessages/SetAllReadWebsiteMessageInputDto.cs [hunk 1]
```diff
@@ -0,0 +1,9 @@
+﻿// Copyright (c) MASA Stack All rights reserved.
+// Licensed under the Apache License. See LICENSE.txt in the project root for license information.
+
+namespace Masa.Mc.Contracts.Admin.Dtos.WebsiteMessages;
+
+public class SetAllReadWebsiteMessageInputDto: GetWebsiteMessageInputDto
```
Comment (MayueCif) [inline src/Contracts/Masa.Mc.Contracts.Admin/Dtos/WebsiteMessages/SetAllReadWebsiteMessageInputDto.cs:6]: ReadAllWebsiteMessageInputDto

File: src/Contracts/Masa.Mc.Contracts.Admin/Dtos/WebsiteMessages/WebsiteMessageChannelListDto.cs [hunk 2]
```diff
@@ -0,0 +1,9 @@
+﻿// Copyright (c) MASA Stack All rights reserved.
+// Licensed under the Apache License. See LICENSE.txt in the project root for license information.
+
+namespace Masa.Mc.Contracts.Admin.Dtos.WebsiteMessages;
+
+public class WebsiteMessageChannelListDto: WebsiteMessageDto
```
Comment (MayueCif) [inline src/Contracts/Masa.Mc.Contracts.Admin/Dtos/WebsiteMessages/WebsiteMessageChannelListDto.cs:6]: WebsiteMessageChannelDto

File: src/Contracts/Masa.Mc.Contracts.Admin/Dtos/MessageTasks/MessageTaskUpsertDto.cs [hunk 3]
```diff
@@ -23,7 +23,7 @@ public class MessageTaskUpsertDto
 
     public MessageTaskSelectReceiverTypes SelectReceiverType { get; set; } = MessageTaskSelectReceiverTypes.ManualSelection;
 
-    public DateTime? SendTime { get; set; }
```
Comment (MayueCif) [inline src/Contracts/Masa.Mc.Contracts.Admin/Dtos/MessageTasks/MessageTaskUpsertDto.cs:26]: [DateTimeOffset](https://docs.microsoft.com/en-us/dotnet/api/system.datetimeoffset) as the default date and time type for application development

File: src/Contracts/Masa.Mc.Contracts.Admin/Dtos/WebsiteMessages/WebsiteMessageDto.cs [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+
+    public string Content { get; set; } = string.Empty;
+
+    public DateTime SendTime { get; set; }
+
+    public bool IsRead { get; set; }
+
+    public DateTime? ReadTime { get; set; }
+
+    public string ZhaiYao { get; set; } = string.Empty;
```
Comment (MayueCif) [inline src/Contracts/Masa.Mc.Contracts.Admin/Dtos/WebsiteMessages/WebsiteMessageDto.cs:24]: ZhaiYao?

File: src/Services/Masa.Mc.Service/Application/WebsiteMessages/Commands/SetAllReadWebsiteMessageCommand.cs [hunk 5]
```diff
@@ -0,0 +1,8 @@
+﻿// Copyright (c) MASA Stack All rights reserved.
+// Licensed under the Apache License. See LICENSE.txt in the project root for license information.
+
+namespace Masa.Mc.Service.Admin.Application.WebsiteMessages.Commands;
+
+public record SetAllReadWebsiteMessageCommand(GetWebsiteMessageInputDto dto) : Command
```
Comment (MayueCif) [inline src/Services/Masa.Mc.Service/Application/WebsiteMessages/Commands/SetAllReadWebsiteMessageCommand.cs:6]: ReadAllWebsiteMessageCommand

File: src/Services/Masa.Mc.Service/Application/WebsiteMessages/Queries/GetPrevWebsiteMessageIdQuery.cs [hunk 6]
```diff
@@ -0,0 +1,9 @@
+﻿// Copyright (c) MASA Stack All rights reserved.
+// Licensed under the Apache License. See LICENSE.txt in the project root for license information.
+
+namespace Masa.Mc.Service.Admin.Application.WebsiteMessages.Queries;
+
+public record GetPrevWebsiteMessageIdQuery(Guid WebsiteMessageId) : Query<Guid>
```
Comment (MayueCif) [inline src/Services/Masa.Mc.Service/Application/WebsiteMessages/Queries/GetPrevWebsiteMessageIdQuery.cs:6]: delete GetPrevWebsiteMessageIdQuery,PrevWebsiteMessageId add WebsiteMessageDto

File: src/Services/Masa.Mc.Service/Application/WebsiteMessages/Queries/GetNextWebsiteMessageIdQuery.cs [hunk 7]
```diff
@@ -0,0 +1,9 @@
+﻿// Copyright (c) MASA Stack All rights reserved.
+// Licensed under the Apache License. See LICENSE.txt in the project root for license information.
+
+namespace Masa.Mc.Service.Admin.Application.WebsiteMessages.Queries;
+
+public record GetNextWebsiteMessageIdQuery(Guid WebsiteMessageId) : Query<Guid>
```
Comment (MayueCif) [inline src/Services/Masa.Mc.Service/Application/WebsiteMessages/Queries/GetNextWebsiteMessageIdQuery.cs:6]: delete

File: src/Services/Masa.Mc.Service/Application/WebsiteMessages/Queries/GetListWebsiteMessageQuery.cs [hunk 8]
```diff
@@ -0,0 +1,9 @@
+﻿// Copyright (c) MASA Stack All rights reserved.
+// Licensed under the Apache License. See LICENSE.txt in the project root for license information.
+
+namespace Masa.Mc.Service.Admin.Application.WebsiteMessages.Queries;
+
+public record GetListWebsiteMessageQuery(GetWebsiteMessageInputDto Input) : Query<PaginatedListDto<WebsiteMessageDto>>
```
Comment (MayueCif) [inline src/Services/Masa.Mc.Service/Application/WebsiteMessages/Queries/GetListWebsiteMessageQuery.cs:6]: GetWebsiteMessageQuery

File: src/Services/Masa.Mc.Service/Domain/WebsiteMessages/Events/WebsiteMessageCreatedDomainEvent.cs [hunk 9]
```diff
@@ -0,0 +1,8 @@
+﻿// Copyright (c) MASA Stack All rights reserved.
+// Licensed under the Apache License. See LICENSE.txt in the project root for license information.
+
+namespace Masa.Mc.Service.Admin.Domain.WebsiteMessages.Events;
+
+public record WebsiteMessageCreatedDomainEvent(Guid UserId, DateTime CheckTime) : DomainEvent
```
Comment (MayueCif) [inline src/Services/Masa.Mc.Service/Domain/WebsiteMessages/Events/WebsiteMessageCreatedDomainEvent.cs:6]: AddWebsiteMessageDomainEvent

File: src/Services/Masa.Mc.Service/Domain/WebsiteMessages/Repositories/IWebsiteMessageRepository.cs [hunk 10]
```diff
@@ -5,5 +5,10 @@ namespace Masa.Mc.Service.Admin.Domain.WebsiteMessages.Repositories;
 
 public interface IWebsiteMessageRepository : IRepository<WebsiteMessage>
 {
-    
+    Task<IQueryable<WebsiteMessage>> GetQueryableAsync();
+    Task<IQueryable<WebsiteMessage>> WithDetailsAsync();
+    Task<WebsiteMessage?> FindAsync(Expression<Func<WebsiteMessage, bool>> predicate, bool include = true, CancellationToken cancellationToken = default(CancellationToken));
+    Task<List<WebsiteMessage>> GetChannelListAsync(Guid userId);
+    Task<WebsiteMessage?> GetPrevWebsiteMessage(Guid id, Expression<Func<WebsiteMessage, bool>> predicate);
```
Comment (MayueCif) [inline src/Services/Masa.Mc.Service/Domain/WebsiteMessages/Repositories/IWebsiteMessageRepository.cs:12]: has difference with FindAsync?

File: src/Services/Masa.Mc.Service/Infrastructure/Repositories/WebsiteMessageRepository.cs [hunk 11]
```diff
@@ -9,4 +9,48 @@ public WebsiteMessageRepository(McDbContext context, IUnitOfWork unitOfWork)
...
+    {
+        var set = Context.Set<WebsiteMessage>().AsNoTracking().Where(x => x.UserId == userId);
+        var sorted = set.OrderByDescending(x => x.CreationTime);
+        var list = set.Select(x => x.ChannelId)
+            .Distinct()
+            .SelectMany(x => sorted.Where(y => y.ChannelId == x).Take(1));
+        return await Task.FromResult(list.OrderByDescending(x => x.CreationTime).ToList());
+    }
+
+    public async Task<WebsiteMessage?> GetPrevWebsiteMessage(Guid id, Expression<Func<WebsiteMessage, bool>> predicate)
```
Comment (MayueCif) [inline src/Services/Masa.Mc.Service/Infrastructure/Repositories/WebsiteMessageRepository.cs:41]: GetPrevWebsiteMessage?

File: src/Web/Masa.Mc.Web.Admin/Components/Messages/MessageDetail.razor.cs [hunk 12]
```diff
@@ -0,0 +1,64 @@
+﻿// Copyright (c) MASA Stack All rights reserved.
+// Licensed under the Apache License. See LICENSE.txt in the project root for license information.
+
+namespace Masa.Mc.Web.Admin.Components.Messages;
+
+public partial class MessageDetail
+{
+    [Parameter]
+    public Guid EntityId { get; set; }
```
Comment (MayueCif) [inline src/Web/Masa.Mc.Web.Admin/Components/Messages/MessageDetail.razor.cs:9]: EntityId?

File: src/Web/Masa.Mc.Web.Admin/Components/Messages/MessageLeft.razor.cs [hunk 13]
```diff
@@ -0,0 +1,39 @@
...
+// Licensed under the Apache License. See LICENSE.txt in the project root for license information.
+
+namespace Masa.Mc.Web.Admin.Components.Messages;
+
+public partial class MessageLeft : AdminCompontentBase
+{
+    [Parameter]
+    public EventCallback<Guid?> OnClick { get; set; }
+
+    private List<WebsiteMessageChannelDto> _entities = new();
```
Comment (MayueCif) [inline src/Web/Masa.Mc.Web.Admin/Components/Messages/MessageLeft.razor.cs:11]: _entities?

File: src/Web/Masa.Mc.Web.Admin/Pages/MessageTemplates/Modules/SmsTemplateCreateModal.razor.cs [hunk 14]
```diff
@@ -14,13 +14,29 @@ public partial class SmsTemplateCreateModal : AdminCompontentBase
...
     private ChannelTypes _channelType;
+    private HubConnection hubConnection;
 
     ChannelService ChannelService => McCaller.ChannelService;
 
     SmsTemplateService SmsTemplateService => McCaller.SmsTemplateService;
 
     MessageTemplateService MessageTemplateService => McCaller.MessageTemplateService;
 
+    protected override async Task OnInitializedAsync()
```
Comment (MayueCif) [inline src/Web/Masa.Mc.Web.Admin/Pages/MessageTemplates/Modules/SmsTemplateCreateModal.razor.cs:25]: repeat code in SmsTemplateEditModal

File: src/Services/Masa.Mc.Service/Domain/WebsiteMessages/Repositories/IWebsiteMessageRepository.cs (hunk 10 above)
Comment (wzh425) [reply src/Services/Masa.Mc.Service/Domain/WebsiteMessages/Repositories/IWebsiteMessageRepository.cs:12]: include

File: src/Web/Masa.Mc.Web.Admin/Store/NoticeState.cs [hunk 15]
```diff
@@ -0,0 +1,44 @@
...
+        }
+    }
+
+    private List<WebsiteMessageDto> _notices = new();
+
+    public delegate Task NoticeChanged();
+
+    public event NoticeChanged? OnNoticeChanged;
+
+
```
Comment (MayueCif) [inline src/Web/Masa.Mc.Web.Admin/Store/NoticeState.cs:29]: delete blank line

File: src/Services/Masa.Mc.Service/Infrastructure/Repositories/WebsiteMessageRepository.cs (hunk 11 above)
Comment (wzh425) [reply src/Services/Masa.Mc.Service/Infrastructure/Repositories/WebsiteMessageRepository.cs:41]: Query the previous message in the station

File: src/Services/Masa.Mc.Service/Domain/WebsiteMessages/Services/WebsiteMessageDomainService.cs [hunk 16]
```diff
@@ -16,6 +16,8 @@ public WebsiteMessageDomainService(IDomainEventBus eventBus, IWebsiteMessageRepo
 
     public virtual async Task CreateAsync(MessageData messageData, MessageTaskHistory taskHistory,MessageReceiverUser item)
     {
+        if (await _messageRecordRepository.FindAsync(x => x.MessageTaskHistoryId == taskHistory.Id && x.UserId == item.UserId) != null) return;
```
Comment (MayueCif) [inline src/Services/Masa.Mc.Service/Domain/WebsiteMessages/Services/WebsiteMessageDomainService.cs:19]: replace any method

**Review by MayueCif — APPROVED** [review]

**Comment by MayueCif** [comment]
@doddgu  LGTM

**Comment by MayueCif** [comment]
@doddgu  LGTM

**Review by doddgu — APPROVED** [review]
LGTM

**Pull request merged**